test/installers/service_installer.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package installers
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/kiegroup/kogito-cloud-operator/pkg/client/meta"
+	"github.com/kiegroup/kogito-cloud-operator/test/framework"
+)
+
+// Map of installed services for namespaces, contains slices of service installers installing services in those namespaces
+var installedServices sync.Map
+
+// ServiceInstaller is the API to install services
+type ServiceInstaller struct {
+	// Install service into specific namespace
+	installIntoNamespace func(string) error
+	// Install service for all namespaces
+	installClusterWide func() error
+	// Return all CRs of this service which exists in this namespace
+	getAllCrsInNamespace func(string) ([]meta.ResourceObject, error)
+	// Uninstall service from specific namespace
+	uninstallFromNamespace func(string) error
+	// Uninstall service from whole cluster
+	uninstallFromCluster func() error
+}
+
+// Install the service into cloud
+func (installer *ServiceInstaller) Install(namespace string) error {
+	// Store service installer for namespace to use for uninstalling purposes
+	if sis, loaded := installedServices.LoadOrStore(namespace, []*ServiceInstaller{installer}); loaded {
```

Comment:
storing/loading before the installation ? 
Could that create any problems when uninstalling it if it failed in-between for some reason ?

Reply:
IMHO even in case of installation failure we should try to uninstall the service, to cleanup the environment.
The possible uninstallation failure should be just logged, shouldn't return an error. Uninstallation shouldn't interrupt the flow or fail the otherwise passing test.